core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -513,9 +513,9 @@ private[log] class Cleaner(val id: Int,
         // note that we will never delete a marker until all the records from that transaction are removed.
         discardBatchRecords = shouldDiscardBatch(batch, transactionMetadata, retainTxnMarkers = retainDeletes)
 
-        // check if the batch contains the last sequence number for the producer. if so, we cannot
+        // check if the batch metadata is currently being cached. if so, we cannot
         // remove the batch just yet or the producer may see an out of sequence error.
-        if (batch.hasProducerId && activeProducers.get(batch.producerId).exists(_.lastSeq == batch.lastSequence))
+        if (batch.hasProducerId && activeProducers.get(batch.producerId).exists(_.hasBatchWithSequenceRange(batch.baseSequence, batch.lastSequence).nonEmpty))
```

Comment:
I agree with you that it is a bit overkill to retain 5 batches in the producer snapshot and also in the log. The cases where we would actually have to rely on them seem quite impossible: you need to a broker to fail, the new leader to be elected, and the leadership to come back to the original broker while the client is still retrying. Alternately, all the brokers may bounce together and lose their in memory state. 

In either case, I think it is fine to return a `DuplicateSequenceException` if we don't have the record metadata. This would mean that the broker side changes are minimal. If we want to cover this case, we can always do so in the future: it is just an optimization. 

I will go ahead and backout the changes to the producer snapshot and retention logic, and add the code to raise a duplicate sequence exception if we he are trying to append a batch with an older sequence but don't have the record metadata in memory.